src/dbvt/mod.rs

```diff
@@ -0,0 +1,1667 @@
+//!     pub aabb: Aabb2<f32>,
+//!     fat_aabb: Aabb2<f32>,
+//! }
+//!
+//! impl Value {
+//!     pub fn new(id: u32, aabb: Aabb2<f32>) -> Self {
+//!         Self {
+//!             index: 0,
+//!             id,
+//!             fat_aabb : aabb.add_margin(Vector2::new(3., 3.)),
+//!             aabb,
+//!         }
+//!     }
+//! }
+//!
+//! impl TreeValue for Value {
+//!     type Vector = Vector2<f32>;
+//!     type Bound = Aabb2<f32>;
+//!
+//!     fn bound(&self) -> &Aabb2<f32> {
+//!         &self.aabb
+//!     }
+//!
+//!     fn fat_bound(&self) -> Aabb2<f32> {
+//!         self.fat_aabb.clone()
+//!     }
+//!
+//!     fn set_index(&mut self, index: usize) {
+//!         self.index = index
+//!     }
+//!
+//!     fn index(&self) -> usize {
+//!         self.index
+//!    }
+//! }
+//!
+//! fn aabb2(minx: f32, miny: f32, maxx: f32, maxy: f32) -> Aabb2<f32> {
+//!    Aabb2::new(Point2::new(minx, miny), Point2::new(maxx, maxy))
+//! }
+//!
+//! fn main() {
+//!     let mut tree = DynamicBoundingVolumeTree::<Value>::new();
+//!     tree.insert(Value::new(10, aabb2(5., 5., 10., 10.)));
+//!     tree.insert(Value::new(11, aabb2(21., 14., 23., 16.)));
+//!     tree.do_refit();
+//!
+//!     let ray = Ray2::new(Point2::new(0., 0.), Vector2::new(-1., -1.).normalize());
+//!     let visitor = ContinuousVisitor::<Ray2<f32>, Value>::new(&ray);
+//!     assert_eq!(0, tree.query(&visitor).len());
+//!
+//!     let ray = Ray2::new(Point2::new(6., 0.), Vector2::new(0., 1.).normalize());
+//!     let visitor = ContinuousVisitor::<Ray2<f32>, Value>::new(&ray);
+//!     let results = tree.query(&visitor);
+//!     assert_eq!(1, results.len());
+//!     assert_eq!(10, results[0].0.id);
+//!     assert_eq!(Point2::new(6., 5.), results[0].1);
+//! }
+//! ```
+//!
+
+pub use self::visitor::*;
+pub use self::util::*;
+
+use std::cmp::max;
+use std::fmt::Debug;
+use std::fmt;
+
+use cgmath::prelude::*;
+use num::NumCast;
+use rand;
+use rand::Rng;
+
+use prelude::*;
+
+mod visitor;
+mod util;
+
+const SURFACE_AREA_IMPROVEMENT_FOR_ROTATION: f32 = 0.3;
+const PERFORM_ROTATION_PERCENTAGE: u32 = 10;
+
+/// Trait that needs to be implemented for any value that is to be used in the
+/// [`DynamicBoundingVolumeTree`](struct.DynamicBoundingVolumeTree.html).
+///
+pub trait TreeValue: Clone + Debug {
+    /// Bounding volume type
+    type Bound;
+
+    /// Vector type used for fat bounds.
+    type Vector: VectorSpace;
+
+    /// Return the bounding volume of the value
+    fn bound(&self) -> &Self::Bound;
+
+    /// Return a fattened bounding volume. For shapes that do not move, this can be the same as the
+    /// base bounding volume. It is recommended for moving shapes to have a larger fat bound, so
+    /// tree rotations don't have to be performed every frame.
+    fn fat_bound(&self) -> Self::Bound;
+
+    /// Used to feed the node index into the value. It is vitally important that this is kept.
+    fn set_index(&mut self, index: usize);
```

Comment:
It doesn't appear to have any semantics within this trait, more like an external hook that you want to store inside, especially since the implementation doesn't really have any freedom in using the index. It just stores it and returns it.

Would doing something like this be appropriate instead?
```rust
values: Vec<(usize, T)>, // store both the index and the value
```

Reply:
I did consider that, and I can't remember exactly why I rejected it back then. I'll see if I can remember.

The trait is there primarily to enable the user to put in any value they like as long as it can return give me a bounding volume. The index might be possible to keep outside, I'll consider it.

Reply:
You're quite correct about the index. I will change to your suggested fix. I had that in the value because at the start of the implementation the nodes could move around. I realized after a while that that would be fairly inefficient, so I added the free list and Node::Nil in the node list "holes", so node indices won't ever change after insertion.

So, by storing the return value from insert, the outside user can record the index if they wish.